src/hierarchies/fa_hierarchy.h

```diff
@@ -0,0 +1,131 @@
+
+#include <google/protobuf/stubs/casts.h>
+
+#include <Eigen/Dense>
+#include <memory>
+#include <vector>
+
+#include "algorithm_state.pb.h"
+#include "base_hierarchy.h"
+#include "hierarchy_id.pb.h"
+#include "hierarchy_prior.pb.h"
+#include "src/utils/distributions.h"
+
+//! Mixture of Factor Analysers hierarchy for multivariate data.
+
+//! This class represents a hierarchical model where data are distributed
+
+namespace FA {
+//! Custom container for State values
+struct State {
+  Eigen::VectorXd mu, psi;
+  Eigen::MatrixXd eta, lambda;
+  Eigen::DiagonalMatrix<double, Eigen::Dynamic> psi_inverse;
+  Eigen::LLT<Eigen::MatrixXd> cov_chol;
+  double prec_logdet;
+};
+
+//! Custom container for Hyperparameters values
+struct Hyperparams {
+  Eigen::VectorXd mutilde, beta;
+  double phi, alpha0;
+  unsigned int q;
+};
+
+};  // namespace FA
+
+class FAHierarchy : public BaseHierarchy<FAHierarchy, FA::State,
+                                         FA::Hyperparams, bayesmix::FAPrior> {
+ public:
+  FAHierarchy() = default;
+  ~FAHierarchy() = default;
+
+  //! Updates hyperparameter values given a vector of cluster states
+  void update_hypers(const std::vector<bayesmix::AlgorithmState::ClusterState>&
+                         states) override;
+
+  //! Updates state values using the given (prior or posterior) hyperparameters
+  FA::State draw(const FA::Hyperparams& params);
+
+  //! Resets summary statistics for this cluster
+  void clear_summary_statistics() override;
+
+  //! Returns the Protobuf ID associated to this class
+  bayesmix::HierarchyId get_id() const override {
+    return bayesmix::HierarchyId::FA;
+  }
+
+  //! Read and set state values from a given Protobuf message
+  void set_state_from_proto(const google::protobuf::Message& state_) override;
+
+  //! Read and set hyperparameter values from a given Protobuf message
+  void set_hypers_from_proto(
+      const google::protobuf::Message& hypers_) override;
+
+  //! Writes current state to a Protobuf message and return a shared_ptr
+  //! New hierarchies have to first modify the field 'oneof val' in the
+  //! AlgoritmState::ClusterState message by adding the appropriate type
+  std::shared_ptr<bayesmix::AlgorithmState::ClusterState> get_state_proto()
+      const override;
+
+  //! Writes current value of hyperparameters to a Protobuf message and
+  //! return a shared_ptr.
+  //! New hierarchies have to first modify the field 'oneof val' in the
+  //! AlgoritmState::HierarchyHypers message by adding the appropriate type
+  std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers> get_hypers_proto()
+      const override;
+
+  //! Returns whether the hierarchy models multivariate data or not
+  bool is_multivariate() const override { return true; }
+
+  //! Saves posterior hyperparameters to the corresponding class member
+  void save_posterior_hypers() {
+    // TODO
+    /* posterior_hypers =
+        static_cast<Derived *>(this)->compute_posterior_hypers();*/
+  }
```

Comment:
What's up with this TODO?

Reply:
My bad, just forgot it in this last review while I had it in the list of things to fix